packages/firestore/src/local/shared_client_state.ts

```diff
@@ -663,9 +659,31 @@ export class WebStorageSharedClientState implements SharedClientState {
     }
   }
 
-  addLocalQueryTarget(targetId: TargetId): void {
+  addLocalQueryTarget(targetId: TargetId): QueryTargetState {
+    let queryState: QueryTargetState = 'not-current';
+
+    // Lookup an existing query state if the target ID was already registered
+    // by another tab
+    if (this.getAllActiveQueryTargets().has(targetId)) {
```

Comment:
I notice there are several places (3-4?) where we call getAllActiveQueryTargets() when we only need to do something simple (like check for the existence of a single targetId).  This worries me a bit since it turns this into an unexpectedly O(n) operation. 

And this particular case could be problematic.  A single client doing _n_ listens will call addLocalQueryTarget() _n_ times and each one will be O(_n_), leading to O(_n_^2) perf overall...

If possible I suggest we find better data structures / plumbing to avoid these hidden O(n) taxes...

Reply:
Added a `isActiveQueryTarget` that I use both here and in SyncEngine. The main difference is that I am no longer accumulating all targets into a new set. It is still not super-efficient, but I would hope that users will not run into this limitation. The number of active clients should be fairly manageable, and managing this directly in a dedicate data structure requires much more state keeping.

`getAllActiveQueryTargets` is now only used from tests.